ipsec/sha1_imp.c

```diff
@@ -0,0 +1,306 @@
+	R(a, b, c, d, e, F4, K4, M(65));
+	R(e, a, b, c, d, F4, K4, M(66));
+	R(d, e, a, b, c, F4, K4, M(67));
+	R(c, d, e, a, b, F4, K4, M(68));
+	R(b, c, d, e, a, F4, K4, M(69));
+	R(a, b, c, d, e, F4, K4, M(70));
+	R(e, a, b, c, d, F4, K4, M(71));
+	R(d, e, a, b, c, F4, K4, M(72));
+	R(c, d, e, a, b, F4, K4, M(73));
+	R(b, c, d, e, a, F4, K4, M(74));
+	R(a, b, c, d, e, F4, K4, M(75));
+	R(e, a, b, c, d, F4, K4, M(76));
+	R(d, e, a, b, c, F4, K4, M(77));
+	R(c, d, e, a, b, F4, K4, M(78));
+	R(b, c, d, e, a, F4, K4, M(79));
+
+	/* Update chaining vars */
+	hd->h0 += a;
+	hd->h1 += b;
+	hd->h2 += c;
+	hd->h3 += d;
+	hd->h4 += e;
+}
+
+
+/* Update the message digest with the contents of INBUF with length INLEN */
+void sha1_write(void *context, uint8_t *inbuf, size_t inlen)
+{
+	sha1_context_t *hd = (sha1_context_t *)context;
+
+	if (hd->count == 64) {
+		sha1_transform(hd, hd->buf);
+		hd->count = 0;
+		hd->nblocks++;
+	}
+
+	if (!inbuf)
+		return;
+
+	if (hd->count) {
+		for (; inlen && hd->count < 64; inlen--)
+			hd->buf[hd->count++] = *inbuf++;
+
+		sha1_write(hd, NULL, 0);
+		if (!inlen)
+			return;
+	}
+
+	while (inlen >= 64) {
+		sha1_transform(hd, inbuf);
+		hd->count = 0;
+		hd->nblocks++;
+		inlen -= 64;
+		inbuf += 64;
+	}
+
+	for (; inlen && hd->count < 64; inlen--)
+		hd->buf[hd->count++] = *inbuf++;
+}
+
+
+/*
+ * The routine final terminates the computation and returns the digest.
+ * The handle is prepared for a new cycle, but adding bytes to the
+ * handle will the destroy the returned buffer.
+ * Returns: 20 bytes representing the digest.
+ */
+void sha1_final(void *context)
+{
+	sha1_context_t *hd = (sha1_context_t *)context;
+	uint32_t t, msb, lsb;
+	uint8_t *p;
+
+	sha1_write(hd, NULL, 0);
+
+	t = hd->nblocks;
+
+	/* multiply by 64 to make a byte count */
+	lsb = t << 6;
+	msb = t >> 26;
+
+	/* add the count */
+	t = lsb;
+	if ((lsb += hd->count) < t)
+		msb++;
+
+	/* multiply by 8 to make a bit count */
+	t = lsb;
+	lsb <<= 3;
+	msb <<= 3;
+	msb |= t >> 29;
+
+	if (hd->count < 56) {
+		hd->buf[hd->count++] = 0x80; /* pad */
+		while (hd->count < 56)
+			hd->buf[hd->count++] = 0; /* pad */
+	}
+
+	/* need one extra block */
+	else {
+		hd->buf[hd->count++] = 0x80; /* pad character */
+		while (hd->count < 64)
+			hd->buf[hd->count++] = 0;
+		sha1_write(hd, NULL, 0); /* flush */
+		;
+		memset(hd->buf, 0, 56); /* fill next block with zeroes */
+	}
+
+	/* append the 64 bit count */
+	hd->buf[56] = msb >> 24;
+	hd->buf[57] = msb >> 16;
+	hd->buf[58] = msb >> 8;
+	hd->buf[59] = msb;
+	hd->buf[60] = lsb >> 24;
+	hd->buf[61] = lsb >> 16;
+	hd->buf[62] = lsb >> 8;
+	hd->buf[63] = lsb;
+	sha1_transform(hd, hd->buf);
+
+	p = hd->buf;
+
+	X(0);
+	X(1);
+	X(2);
+	X(3);
+	X(4);
+}
+
+
+/*
+ * Shortcut functions which puts the hash value of the supplied buffer
+ * into outbuf which must have a size of 20 bytes.
+ */
+void sha1_hash(char *outbuf, const char *buffer, size_t length)
+{
+	/* (PROBLEM) */
+	static sha1_context_t hd;
```

Comment:
I guess it caused stack overflow... But please take a note that this is now not thread-safe. Maybe move hd back to stack?